### ISSUE TITLE
Help with Training MATGL Force Field on HSE Data – Force Predictions Are Off

### ISSUE THREAD

**Issue by msehabibur** [materialsvirtuallab/matgl#566]
Hi,

I’ve been working on training a MATGL force field model from scratch using HSE-level DFT data. While the energy predictions are excellent, I'm encountering significant discrepancies in the force predictions, which is causing issues during the structural optimization of crystals.

I’ve attached my train.py script below. I’d really appreciate it if you could take a look and suggest any improvements or adjustments that might help improve the force accuracy.

Thank you so much for your time and help!

from __future__ import annotations

import os
os.environ['LD_LIBRARY_PATH'] = '/apps/gilbreth/cuda-toolkit/cuda-11.2.0/lib64:' + os.environ.get('LD_LIBRARY_PATH', '')
import json
import dgl
import shutil
import warnings
import logging
import numpy as np
import pandas as pd
import ast
from pymatgen.core.structure import Structure
import torch
import pytorch_lightning as pl
from functools import partial
from pytorch_lightning.loggers import CSVLogger

# Import updated MGLDataset, MGLDataLoader, and collate_fn_pes
from mgl_data_utils import MGLDataset, MGLDataLoader, collate_fn_pes

import matgl
from matgl.ext.pymatgen import Structure2Graph
from matgl.models import M3GNet
from matgl.utils.training import PotentialLightningModule
from matgl.config import DEFAULT_ELEMENTS

import argparse

# Parse command-line arguments for checkpoint resumption
parser = argparse.ArgumentParser(description='Resume M3GNet Training')
parser.add_argument('--checkpoint', type=str, default=None, help='Path to checkpoint file to resume training')
args = parser.parse_args()

print(dgl.__version__)
# Set up logging to file and console
log_file = "training.log"
logging.basicConfig(level=logging.INFO, format='%(asctime)s - %(levelname)s - %(message)s', handlers=[
    logging.FileHandler(log_file),
    logging.StreamHandler()
])
logger = logging.getLogger(__name__)

logger.info("Script started...")

# To use GPU
torch.set_default_device("cuda")
logger.info(f"Is CUDA available: {torch.cuda.is_available()}")

# Set CUDA as the default tensor type
torch.set_default_dtype(torch.float32)
torch.set_float32_matmul_precision('medium')

# To suppress warnings for clearer output
warnings.simplefilter("ignore")

# Parse forces and stresses as arrays

# Load only the first 10 rows
df = pd.read_csv('merged_data.csv')

# Parse the Forces and Stresses columns as arrays using json.loads
df['Forces_array'] = df['Forces'].apply(lambda x: np.array(ast.literal_eval(x)))
df['Stresses_array'] = df['Stress'].apply(lambda x: np.array(ast.literal_eval(x)) * -0.1)

# Parse the Structure column into pymatgen Structure objects
structures = [Structure.from_dict(json.loads(s)) for s in df['Structure']]

# Prepare energy, forces, and stress labels
energies = df['Energy'].tolist()
forces = [f.tolist() for f in df['Forces_array']]
stresses = [s.tolist() for s in df['Stresses_array']]

labels = {
    "energies": energies,
    "forces": forces,
    "stresses": stresses,
}

element_types = DEFAULT_ELEMENTS
converter = Structure2Graph(element_types=element_types, cutoff=6.0)

save_dir = './'
os.makedirs(save_dir, exist_ok=True)

dataset = MGLDataset(threebody_cutoff=6.0, structures=structures, converter=converter, labels=labels, include_line_graph=True, save_dir=save_dir)

train_data, val_data, test_data = dgl.data.utils.split_dataset(
    dataset,
    frac_list=[0.85, 0.1, 0.05],
    shuffle=True,
    random_state=42,
)

# if you are not intended to use stress for training, switch include_stress=False!
my_collate_fn = partial(collate_fn_pes, include_line_graph=True, include_stress=True)

# Create CUDA generator
generator = torch.Generator(device="cuda")
generator.manual_seed(42)
logger.info(f"Generator device: {generator.device}")

train_loader, val_loader, test_loader = MGLDataLoader(
    train_data=train_data,
    val_data=val_data,
    test_data=test_data,
    collate_fn=my_collate_fn,
    batch_size=16,
    num_workers=0,
    generator=generator,
)

# Check generator in DataLoader
logger.info(f"Train DataLoader generator device: {train_loader.generator.device}")
logger.info(f"Validation DataLoader generator device: {val_loader.generator.device}")

# download a pre-trained M3GNet
#m3gnet_nnp = matgl.load_model("M3GNet-MP-2021.2.8-PES")
#model_pretrained = m3gnet_nnp.model
#property_offset = m3gnet_nnp.element_refs.property_offset

# Define the M3GNet model
model = M3GNet(
    element_types=element_types,
    is_intensive=False,
)

#lit_module_finetune = PotentialLightningModule(model=model, lr=5e-4, include_line_graph=True, stress_weight=0.0001)
##lit_module_finetune = PotentialLightningModule(model=model, lr=5e-4, include_line_graph=True, stress_weight=0.00)
lit_module_finetune = PotentialLightningModule(model=model, lr=5e-4, include_line_graph=True, stress_weight=0.01)

# If you wish to disable GPU or MPS (M1 mac) training, use the accelerator="cpu" kwarg.
logger.info("Setting up trainer...")
logger_csv = CSVLogger("logs", name="M3GNet_finetuning")
trainer = pl.Trainer(max_epochs=1200, accelerator="gpu", devices=1, logger=logger_csv, inference_mode=False)
logger.info("Starting training...")
if args.checkpoint:
    logger.info(f"Resuming training from checkpoint: {args.checkpoint}")
    trainer.fit(model=lit_module_finetune, train_dataloaders=train_loader, val_dataloaders=val_loader, ckpt_path=args.checkpoint)
else:
    logger.info("Starting training from scratch.")
    trainer.fit(model=lit_module_finetune, train_dataloaders=train_loader, val_dataloaders=val_loader)
trainer.fit(model=lit_module_finetune, train_dataloaders=train_loader, val_dataloaders=val_loader)

# save trained model
model_save_path = "./"
lit_module_finetune.model.save(model_save_path)
logger.info(f"Model saved to {model_save_path}")

**Comment by msehabibur** [comment]
This is my force parity plot 

![Image](https://github.com/user-attachments/assets/460742f1-3828-4dcb-b965-c8708ecea119)

**Comment by msehabibur** [comment]
@shyuep

**Comment by kenko911** [comment]
Hi @msehabibur,

Thank you for sharing the training script—it looks good to me overall. However, I’m quite puzzled by the force parity plot. Given the reported MAE of approximately 0.02 eV/Å, it appears inconsistent with both the reported r^2 score and the parity plot itself. I recommend thoroughly reviewing each step of your analysis process to ensure accuracy before reporting any issues.
